src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs

```diff
@@ -231,6 +231,26 @@ private CompletionProvider GetProviderByName(string providerName)
                     break;
             }
 
+            // Phase 1: Completion Providers decide if they are triggered based on textual analysis
+            // Phase 2: Completion Providers use syntax to confirm they are triggered, or decide they are not actually triggered and should become an augmenting provider
+            // Phase 3: Triggered Providers are asked for items
+            // Phase 4: If any items were provided, all augmenting providers are asked for items
+            // This allows a provider to be textually triggered but later decide to be an augmenting provider based on deeper syntactic analysis.
+
+            var additionalAugmentingProviders = new List<CompletionProvider>();
+            foreach (var provider in triggeredProviders)
+            {
+                if (trigger.Kind == CompletionTriggerKind.Insertion)
+                {
+                    if (!await provider.IsSyntacticTriggerCharacterAsync(document, caretPosition, trigger, options, cancellationToken).ConfigureAwait(false))
```

Comment:
nit: consider merging if statements.

Reply:
Consider this considered, but I think I'll leave it in this case.